### PR TITLE
Add host.containers.internal entry into container's etc/hosts

### LINKED ISSUE #5651: hostname for container gateway
/kind feature

**Description**

Provide a dns entry to containers pointing to their gateway.

Context : as a developer, one of my setup involves a service running on the host machine, that should be proxied by nginx running in a container.

**Describe the results you received:**

I could make it work by writing the container's static gateway ip in nginx's config.

**Describe the results you expected:**

It would have been nice to have a standard hostname for the gateway (I believe docker has `host.docker.internal` for that). Or did I miss some easy alternative ?

**Output of `podman version`:**

```
Version:            1.8.2
RemoteAPI Version:  1
Go Version:         go1.14
Git Commit:         028e3317eb1494b9b2acba4a0a295df80fae66cc
Built:              Sat Mar 21 15:30:34 2020
OS/Arch:            linux/amd64```

### PR DESCRIPTION
**Pull request by bblenard**
This change adds the entry `host.containers.internal` to the `/etc/hosts`
file within a new containers filesystem. The ip address is determined by
the containers networking configuration and points to the gateway address
for the containers networking namespace.

Closes #5651

Signed-off-by: Baron Lenardson <lenardson.baron@gmail.com>

### CODE REVIEW

File: libpod/container_internal_linux.go
Comment (mheon) [inline libpod/container_internal_linux.go:1404]: Can drop this block and make `depCtr, err = c.getRootNetNsDepCtr()` a `:=`

Will be a little cleaner

File: libpod/networking_linux.go
Comment (mheon) [inline libpod/networking_linux.go:39]: Can you keep this block `const` and then add a separate `var` block below for the slirp4netns bit?

**Comment by mheon** [comment]
I'd like @AkihiroSuda or someone else familiar with slirp to verify the slirp address calculation bits. Overall looks good, few small comments.

**Comment by TomSweeneyRedHat** [comment]
Nothing over @mheon's comments.  @giuseppe if you have a spare eyeball for this, that would be great.

File: libpod/container_internal_linux.go
Comment (bblenard) [reply libpod/container_internal_linux.go:1345]: I don't think this can be null unless there is an error, although I could be wrong. So would you like a note about it returning an nil on error?

File: test/system/500-networking.bats
Comment (AkihiroSuda) [inline test/system/500-networking.bats:102]: I'm not feeling this behavior is correct.
The CIDR+2 address isn't useful when slirp4netns is launched with `--disable-host-loopback`.
And `--disable-host-loopback` should not be removed, as exposing bare 127.0.0.1 to containers might result in container-breakout.

I suggest exposing the "real" host IP regardless to rootful/rootless.

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]

File: test/system/500-networking.bats
Comment (bblenard) [reply test/system/500-networking.bats:102]: So what would be the alternative? Are you suggesting to derive all interface ip addresses not including `lo`?

The issue (#5651) was to add behavior that reflects docker's `host.docker.internal`. 

I had a look through some of the docker code and as far as I can tell that is a configuration option given to the daemon so if a user asks for the host gateway address to be added to their container it is added based on that configuration [docker-ce reference](https://github.com/docker/docker-ce/blob/cc71e4fd9a9933078d79daaf69c8422c1bfa1618/components/engine/cmd/dockerd/config.go#L67)

So should podman have a similar configuration option?

Comment (AkihiroSuda) [reply test/system/500-networking.bats:102]: > Are you suggesting to derive all interface ip addresses not including lo?

The first interface (eth0) would be fine

File: libpod/container_internal_linux.go
Comment (rhatdan) [inline libpod/container_internal_linux.go:1806]: This last block is not necessary.  depCtr will be nil by default.

File: libpod/container_internal_linux.go
Comment (rhatdan) [inline libpod/container_internal_linux.go:1820]: Is this unexpected, or would this just be the situation if I run a container with `podman run --net=none ...`

IE Is this worth logging?

**Comment by rhatdan** [comment]
@bblenard Thanks for working on this.

File: test/system/500-networking.bats
Comment (bblenard) [reply test/system/500-networking.bats:102]: I'm not sure about this solution.

There can be so many different interfaces in so many different configurations that trying to derive an ip address from one of the interfaces seems unstable. I don't necessarily disagree that exposing the loopback interface to the container could result in a container breakout but this change doesn't give that access it simply puts a `/etc/hosts` entry for that interface.

Like I said above. I think the only viable alternative is to make the interface a configurable option similar to dockerd's `--host-gateway-ip` launch option.

File: libpod/container_internal_linux.go
Comment (bblenard) [reply libpod/container_internal_linux.go:1820]: This would be the case for `podman run --net=none` so I suppose it isn't that unexpected. Should I just quietly continue or change up the debug message?

File: libpod/container_internal_linux.go
Comment (bblenard) [reply libpod/container_internal_linux.go:1806]: I'm not sure syntactically how to get rid of that block :grimacing: 

I added it because I wanted to ensure mutual exclusion between shared container network namespace configurations and containers with network namespaces created for them. Basically I didn't want the situation where

```
        var depCtr *Container
	if c.config.NetNsCtr != "" { // **TRUE**
		// ignoring the error because there isn't anything to do
		depCtr, _ = c.getRootNetNsDepCtr()
	} 
	if len(c.state.NetworkStatus) != 0 { // **TRUE**
		depCtr = c
	}
```

But If that is not possible / something to worry about I can make them both `if`'s and remove the unnecessary block

File: libpod/networking_linux.go
Comment (bblenard) [reply libpod/networking_linux.go:39]: I have changed this I just don't want to kick off another CICD run until I get the other things sorted out.

File: libpod/container_internal_linux.go
Comment (bblenard) [reply libpod/container_internal_linux.go:1404]: https://github.com/containers/podman/pull/9972#discussion_r615246254

File: libpod/container_internal_linux.go
Comment (bblenard) [reply libpod/container_internal_linux.go:1820]: @rhatdan bump just in case you didn't see my response

File: libpod/container_internal_linux.go
Comment (bblenard) [reply libpod/container_internal_linux.go:1806]: @rhatdan just in case you didn't see my response.

File: test/system/500-networking.bats
Comment (bblenard) [reply test/system/500-networking.bats:102]: @AkihiroSuda bump in case you didn't see my last response

File: libpod/container_internal_linux.go
Comment (rhatdan) [reply libpod/container_internal_linux.go:1806]: Yes, that is fine.

**Comment by TomSweeneyRedHat** [comment]
@rhatdan can you merge this if it LGTY?  I think there was one last "is this OK" type of question for you in the review comments on this one.

File: test/system/500-networking.bats
Comment (AkihiroSuda) [reply test/system/500-networking.bats:102]: Having an option SGTM

**Comment by AkihiroSuda** [comment]
Bikeshedding:  For consistency with `host.docker.internal`, the corresponding value for Podman should be `host.podman.internal` or `host.containers.internal` rather than `host.gateway.internal`?

**Comment by rhatdan** [comment]
Yes lets go with `host.containers.internal` and then we can merge.

**Comment by bblenard** [comment]
I just pushed up some code and assuming the tests all pass there are just a few small things I want to make sure everyone is on the same page about:

1) [this thread](https://github.com/containers/podman/pull/9972#discussion_r610340456). @AkihiroSuda mentioned their concerns about using the slirpns gateway address, we discussed a bit and contemplated using a configuration option to specify an ip address for the `host.containers.internal` address. @rhatdan / @mheon / @TomSweeneyRedHat thoughts? If that is the route we want to go I'll have to rework most of this code :)

2) @rhatdan I wasn't sure what ["Yes, that is fine."](https://github.com/containers/podman/pull/9972#discussion_r620705065) meant exactly so I left the code as it was. I can change it if I mis-interpreted your response though.

**Comment by AkihiroSuda** [comment]
The commit message ( host.gateway.internal ) seems to need to be updated

**Comment by bblenard** [comment]
> The commit message ( host.gateway.internal ) seems to need to be updated

Fixed f5e7b40d66646b3e6517d989d8078110172808de

**Comment by rhatdan** [comment]
LGTM

**Comment by rhatdan** [comment]
@containers/podman-maintainers PTAL

File: libpod/networking_slirp4netns.go
Comment (Luap99) [inline libpod/networking_slirp4netns.go:337]: I don't think we should use globals, this will fail when used with the podman service.

File: libpod/networking_slirp4netns.go
Comment (Luap99) [inline libpod/networking_slirp4netns.go:318]: This block has to above `if  havePortMapping...`

File: libpod/networking_slirp4netns.go
Comment (bblenard) [reply libpod/networking_slirp4netns.go:337]: Okay, I'll look at moving away from the globals. I think when I initially wrote this those were constant values that I changed to variables so I didn't even think about if they **should** be globals. Thanks for catching that :)

File: libpod/networking_slirp4netns.go
Comment (bblenard) [reply libpod/networking_slirp4netns.go:318]: I'll fix that once I figure out how to move away from globals like you suggested above.

File: libpod/networking_slirp4netns.go
Comment (Luap99) [reply libpod/networking_slirp4netns.go:337]: I think you can add a new field (maybe called slirp4netnsSubnet) to the container struct here https://github.com/containers/podman/blob/663ea964733f5d7ccef9b457b33e97ebdec79e70/libpod/container.go#L97
and add methods like GetSlirp4netnsIP, GetSlirp4netnsGateway and GetSlirp4netnsDNS which returns the correct ip calculated from the stored subent.

Comment (bblenard) [reply libpod/networking_slirp4netns.go]: Fixed: https://github.com/containers/podman/commit/3ce696e05426bc03a541ba6f36f8ee48080c0c66#diff-b50e7e875575fcd700a1431a314b378fc179af9a4d1da93cbb62a42d10949c0aR335

I made the helper functions generatl libpod functions that take subnets instead of Container functions that reference internal struct members because I didn't think it was a container specific thing. I also essentially reverted your changes here: https://github.com/containers/podman/commit/f99b7a314bf6d285b92a528d1a0800c35dfdc603 since I had a merge conflict and my changes covered what you did with that commit as far as I could tell.

File: libpod/networking_slirp4netns.go
Comment (bblenard) [reply libpod/networking_slirp4netns.go]: Fixed: c0335e77819e23be94bb55d883e18bdd6b742195

**Review by Luap99 — COMMENTED** [review]
Thanks this looks good. One small nit.

File: libpod/container_internal_linux.go
Comment (rhatdan) [inline libpod/container_internal_linux.go:1802]: This should be a warning not an error,  If the user can not do anything about it, then it should just warn.

**Comment by Luap99** [comment]
@edsantiago PTAL at the system tests.

File: libpod/container_internal_linux.go
Comment (rhatdan) [inline libpod/container_internal_linux.go:1716]: Warning not Error.

File: libpod/container_internal_linux.go
Comment (rhatdan) [inline libpod/container_internal_linux.go:1850]: Warning not error.

**Comment by rhatdan** [comment]
LGTM once Error->Warnin

File: test/system/500-networking.bats
Comment (edsantiago) [inline test/system/500-networking.bats:166]: Just curious: what problem does `random_rfc1918_subnet()` have that prevents you from using it?

Comment (bblenard) [reply test/system/500-networking.bats]: I wasn't aware of that function. I added it to the tests I wrote :+1:

File: libpod/networking_linux.go
Comment (bblenard) [reply libpod/networking_linux.go:363]: Fixed: ef6027b2db9d7ebd4a294af002c08c0c93d0e5db

File: libpod/container_internal_linux.go
Comment (bblenard) [reply libpod/container_internal_linux.go]: Fixed: ef6027b2db9d7ebd4a294af002c08c0c93d0e5db

File: libpod/container_internal_linux.go
Comment (bblenard) [reply libpod/container_internal_linux.go]: Fixed: ef6027b2db9d7ebd4a294af002c08c0c93d0e5db

File: libpod/container_internal_linux.go
Comment (bblenard) [reply libpod/container_internal_linux.go]: Fixed: ef6027b2db9d7ebd4a294af002c08c0c93d0e5db

**Review by Luap99 — APPROVED** [review]
LGTM

File: test/system/500-networking.bats
Comment (edsantiago) [reply test/system/500-networking.bats]: Thanks! One suggestion, though: I find string concatenation easier to read than stripping. Would you consider:
```
    subnet=$(random_rfc1918_subnet)
    ... :cidr="$subnet.0/24"
    .... grep "$subnet"
    ... nameserger "${subnet}.3"
```
That removes the need for all those complicated fragile `%.*`, and makes life much easier for future maintainers.

Comment (bblenard) [reply test/system/500-networking.bats]: Sounds good. Fixed: c8dfcce6db0adfa2f91d979271e8c2725eb753a6

File: test/system/500-networking.bats
Comment (edsantiago) [inline test/system/500-networking.bats:166]: Much better, thank you, but I'm kind of uncomfortable now with the variable name: a CIDR is defined as a full subnet, slash, and mask. A future maintainer may find this a stumbling point, and waste time (e.g. on line 169) wondering how a CIDR can have `.2` appended. This is about a 3-out-of-10 gripe, I'm not going to insist on you changing it, but I'd like to bring it to the attention of others in case someone else on the team has a strong opinion either way.

**Comment by rhatdan** [comment]
Thanks @bblenard Nice work.

**Comment by bblenard** [comment]
> Thanks @bblenard Nice work.

Thanks! I've been enjoying working on podman :)

**Pull request merged**
